### PR TITLE
Ducdev/final version

### PR DESCRIPTION
**Pull request by ducdev**
```
Added:
 - Styling the UI for both web and mobile follow provided designs.
Refactored
 - I removed promises such as some `this.setState` which is nonsense
 - Fat arrow functions is preferred follow ES6
Fixed:
 - `KI-01`: onChange handler should be debounced properly to avoid multiple requests.
 - `KI-02`: add a loading indicator to help users acknowledging the progress.
```

### CODE REVIEW

**Comment by ducdev** [comment]
Hey @longtran-pops the only concern for now is `lodash`, please let me know if I should make changes. :)

**Comment by longtran-pops** [comment]
> Hey @longtran-pops the only concern for now is `lodash`, please let me know if I should make changes. :)

In order to see how you solve the problem, I would like you to use your own solution instead. Even though you copy a part of `lodash`'s code, it is still fine.

**Comment by ducdev** [comment]
Hi @longtran-pops I've made changes, please help reviewing.

File: src/App.js [hunk 1]
```diff
@@ -1,55 +1,63 @@
...
+    // Get all the items which start with `keyword`
+
+    this.setState({ loading: true })
+
+    const results = await FakeSearchAPI.search(this.state.keyword)
+
+    // Update suggestion list
+
+    if (results && Array.isArray(results)) {
+      this.setState({ loading: false })
```
Comment (longtran-pops) [inline src/App.js]: You only set loading when there is `results` and the `results` is an Array. Mean that if the `results` is not an Array or is null, then it still remain loading

File: src/components/widgets/PAutoComplete.js [hunk 2]
```diff
@@ -1,26 +1,39 @@
...
+const capitalizeFirstLetter = string => string.charAt(0).toUpperCase() + string.slice(1)
+
 export default (props) => {
   return (
-    <div>
+    <div
+      className="auto-complete"
+    >
       <PInput
+        disabled={props.loading}
```
Comment (longtran-pops) [inline src/components/widgets/PAutoComplete.js]: Disable input while loading can prevent unexpected load but it hurts the user experience.

File: src/components/widgets/PAutoComplete.js [hunk 3]
```diff
@@ -1,26 +1,39 @@
...
               key={item}
               onPress={() => {
                 props.onSelect && props.onSelect(item)
               }}
-            >
-              {item}
-            </PListItem>
+              text={{
+                highlighted: props.value,
+                normal: item.replace(capitalizeFirstLetter(props.value), ''),
```
Comment (longtran-pops) [inline src/components/widgets/PAutoComplete.js]: Change inside text from ChildNode to Attribute would not be a good approach to highlight the text.
Use ChildNode can open a gate to use any thing as an inner text such as a button, label or a complex elements.

**Review by longtran-pops — CHANGES_REQUESTED** [review]

**Comment by ducdev** [comment]
Hi @longtran-pops I totally agree with your feedbacks, I've updated my PR. Please review it again. Thank you.

File: src/App.js (hunk 1 above)
Comment (ducdev) [reply src/App.js]: Yes, the loading status should be revoked before the if condition, my bad.

File: src/components/widgets/PAutoComplete.js (hunk 2 above)
Comment (ducdev) [reply src/components/widgets/PAutoComplete.js]: I already applied another technique which is silent totally and won't hurt the UX anymore. Hope you like it.

File: src/components/widgets/PAutoComplete.js (hunk 3 above)
Comment (ducdev) [reply src/components/widgets/PAutoComplete.js]: I can't agree more!

**Comment by longtran-pops** [comment]
Hi @ducdev,

Sorry for late reply, I have just gone through your solution once again. Even though you have fixed some issues and done some refactoring, It was still under my expectations.

There're some open points:
- Using `debounce` only won't fix the issue `KI-01`, it only avoids some cases.
- Using loading indicator just let the user know we are fetching data. I would like to see how can we make it **look** (just a feeling) faster.

And it is also good to have:
- UnitTest for some stuffs you added in (such as `debounce`)
- Make the code cleaner to re-use, extends.

But after all, you're still really good at collaboration via git. I'm really happy about that. Hope you can improve your technical skills and come to us next time.

Thanks

**Comment by ducdev** [comment]
Hi @longtran-pops thank you!

**Pull request closed without merging**